### ISSUE TITLE
Support declarative HTTP Service clients

### ISSUE THREAD

**Issue by mhalbritter** [spring-projects/spring-boot#31337]
Spring Framework 6.0 [introduces declarative HTTP clients](https://docs.spring.io/spring-framework/docs/6.0.0-SNAPSHOT/reference/html/integration.html#rest-http-interface). We should add some auto-configuration in Boot which supplies for example the `HttpServiceProxyFactory`.

We could even think about an annotation with which the HTTP interface can be annotated and then directly injected into a consumer (like `@FeignClient`).

**Comment by livk-cloud** [comment]
> I think this one can try to inject IOC by myself
> For example I do

```java
public class HttpServiceFactory implements BeanFactoryAware, ImportBeanDefinitionRegistrar, ResourceLoaderAware {

	private final HttpServiceProxyFactory proxyFactory;

	private BeanFactory beanFactory;

	private ResourceLoader resourceLoader;

	public HttpServiceFactory() {
		WebClient client = WebClient.builder().build();
		this.proxyFactory = HttpServiceProxyFactory.builder(new WebClientAdapter(client)).build();
	}

	@Override
	public void registerBeanDefinitions(@NonNull AnnotationMetadata importingClassMetadata,
			@NonNull BeanDefinitionRegistry registry) {
		List<String> packages = AutoConfigurationPackages.get(this.beanFactory);
		Set<Class<?>> typesAnnotatedClass = SpringUtils.findByAnnotationType(HttpExchange.class, resourceLoader,
				packages.toArray(String[]::new));
		for (Class<?> exchangeClass : typesAnnotatedClass) {
			BeanName name = AnnotationUtils.getAnnotation(exchangeClass, BeanName.class);
			String beanName = name != null ? name.value()
					: CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_CAMEL, exchangeClass.getSimpleName());
			registry.registerBeanDefinition(beanName, getBeanDefinition(exchangeClass));
		}
	}

	private <T> BeanDefinition getBeanDefinition(Class<T> exchangeClass) {
		return new RootBeanDefinition(exchangeClass, () -> proxyFactory.createClient(exchangeClass));
	}

	@Override
	public void setResourceLoader(@NonNull ResourceLoader resourceLoader) {
		this.resourceLoader = resourceLoader;
	}

	@Override
	public void setBeanFactory(@NonNull BeanFactory beanFactory) throws BeansException {
		this.beanFactory = beanFactory;
	}

}
```

**Comment by spencergibb** [comment]
What would an additional annotation bring beyond `@HttpMapping`

**Comment by rstoyanchev** [comment]
The underlying client may need to be configured with different base URL, codecs, etc. That means detecting `@HttpExchange` annotated interfaces and declaring beans for them could become inflexible if it's based on the same underlying client setup. 

Given an `HttpServiceProxyFactory`, it's trivial to create a proxy, so don't see a lot of gain from automating that. What I am thinking about though is that it should be possible for Boot to create and auto-configure one `HttpServiceProxyFactory` instance that can then be combined that with different client setups.

Currently `HttpServiceProxyFactory` takes `HttpClientAdapter` in the constructor, but we could make a change to allow it to be passed to an overloaded `createClient(Class<?>, HttpClientAdapter)` method. So you could inject the same `HttpServiceProxyFactory` anywhere, and either use it with the default client setup (based on `WebClientCustomizer` and `WebClient.Builder`), or optionally, also inject `WebClient.Builder` and use a client setup that deviates from the default.

**Comment by rstoyanchev** [comment]
Thanks, this helps me to move my thought process forward. I see now it is necessary to separate more formally the `HttpServiceProxyFactory` from the underlying client.

I've an experiment locally where `HttpServiceProxyFactory` expects the `HttpClientAdapter` to be passed in every time `createClient` is called. Separately, there is a `WebClientServiceProxyFactory` that is created with an `HttpServiceProxyFactory` and a `WebClient` and exposes a `createClient` with just the proxy interface. 

The Boot auto-config could then declare a single `HttpServiceProxyFactory` bean, and applications would create any number of `WebClientServiceProxyFactory` beans, each delegating to the same `HttpServiceProxyFactory` and the `WebClient` configured for a specific remote.

**Comment by rstoyanchev** [comment]
After a few different experiments, I think trying to have one `HttpServiceProxyFactory` for many client instances brings additional complexity with little gain. The easiest to understand model remains, one `HttpServiceProxyFactory` for one client. It's reasonably simple even without any help from Boot:

```java
@Bean
HttpServiceProxyFactory httpServiceProxyFactory1(WebClient.Builder clientBuilder) {
	WebClient client = clientBuilder.baseUrl("http://host1.com").build();
	return new HttpServiceProxyFactory(new WebClientAdapter(client));
}

@Bean
HttpServiceProxyFactory httpServiceProxyFactory2(WebClient.Builder clientBuilder) {
	WebClient client = clientBuilder.baseUrl("http://host2.com").build();
	return new HttpServiceProxyFactory(new WebClientAdapter(client));
}
```

A couple of extra shortcuts on `WebClientAdapter` could make this a one-liner:
```java
@Bean
HttpServiceProxyFactory httpServiceProxyFactory1(WebClient.Builder clientBuilder) {
	return WebClientAdapter.createProxyFactory(clientBuilder.baseUrl("http://host1.com"));
}

@Bean
HttpServiceProxyFactory httpServiceProxyFactory2(WebClient.Builder clientBuilder) {
	return WebClientAdapter.createProxyFactory(clientBuilder.baseUrl("http://host2.com"));
}
```

If we settle on the above as the expected configuration, then I think it's not essential to have any Boot auto-config to start, although some ideas may still come along. Perhaps, specifying baseUrl's in properties, which would allow Boot to create the above beans?

**Comment by wilkinsona** [comment]
Thanks, Rossen.

> Perhaps, specifying baseUrl's in properties, which would allow Boot to create the above beans?

We don't yet support specifying multiple property values to auto-configure multiple beans anywhere in Boot. It's something that we'd like to do, but it's a complex and wide-ranging topic. https://github.com/spring-projects/spring-boot/issues/15732 is tracking auto-configured multiple DataSources, for example.

Having discussed this today, we don't think there's anything to do in Boot at this time. We can revisit this if the picture changes for auto-configuring multiple beans.

**Comment by rstoyanchev** [comment]
Note that there is now https://github.com/spring-projects/spring-framework/issues/29296, which will likely give us a better model for dealing with multiple `HttpServiceProxyFactory` instances for different remotes.

**Comment by bclozel** [comment]
Reopening because of changes made in spring-projects/spring-framework#29296
Spring Boot could contribute a pre-configured `HttpServiceProxyFactory.Builder` to the context so developer can build their own client from it.

**Comment by hannah23280** [comment]
With regard to this nice tutorial on HTTP Interface https://softice.dev/posts/introduction_to_spring_framework_6_http_interfaces/,

I don't quite understand. Why do developer need to manually write a @Bean method that will return the proxy bean (which implement the interface) especially if we are using spring boot? 
I recall using @FeignClient, I do not have to define any proxy bean for it, so I presume spring boot will do for us.

Also why would one use Http Interface over @FeignClient ?

**Comment by DanielLiu1123** [comment]
> We could even think about an annotation with which the HTTP interface can be annotated and then directly injected into a consumer (like `@FeignClient`).

I think we need an annotation like `@EnableFeignClients` rather than `@FeignClient`.

We can already know whether an interface is a http client through `@HttpExchange`, we need an annotation to scan the interfaces and register beans (like `@EnableFeignClients`).

**Comment by alsikorski** [comment]
I think people have gotten used to using the Feign client approach.

Here you can find very similar aproach: [Exchange client](https://github.com/MelonProjectCom/exchange-client)

It is really simple to use.

**Comment by maciejwalkowiak** [comment]
I have prototyped following approach, that reduces the boilerplate to minimum:

### `@HttpClient` annotation to mark interface as an http client and add option to set the `WebClient` bean name to use.

```java
@HttpClient("todo-client")
public interface TodoClient {
    @GetExchange("/todos")
    List<Todo> get();
}
```

This annotation is processed by an registrar implementing `ImportBeanDefinitionRegistrar` that registers bean definition for each http client

### Creating `WebClient` instances from the environment

Considering that many web clients are relatively simple, there is a common set of properties that can be set with simple properties: url, basic auth, timeouts etc. 

Given this, there is an option to create WebClients through yaml/properties like this:

```yaml
http.clients:
    todo-client:
        url: https://jsonplaceholder.typicode.com
    bar:
        url: http://foo/bar
```

If you believe it makes sense I can prepare a PR or if it's too early to say I can release it as a separate project that will get deprecated once Spring Boot has similar functionality built in.

**Comment by OlgaMaciaszek** [comment]
Another proposed implementation: https://github.com/joshlong/declarative-client-registration by @joshlong .

**Comment by DanielLiu1123** [comment]
Let me introduce once again to [httpexchange-spring-boot-starter](https://github.com/DanielLiu1123/httpexchange-spring-boot-starter), which is probably the most comprehensive implementation I could find.

This project is entirely configuration-driven and can achieve the same functionalities as Spring Cloud OpenFeign without introducing any external annotations. Including setting different timeout durations for each client, integration with Spring Cloud LoadBalancer, dynamic refresh, and more.

```yaml
http-exchange:
  base-packages: [com.example.api]
  connect-timeout: 1000
  read-timeout: 3000
  client-type: rest_client
  channels:
    - base-url: http://order
      clients:
        - com.example.order.api.*
    - base-url: http://user
      read-timeout: 5000
      classes:
        - com.example.user.api.*
````

The main goals of this project:
- Promote the use of `@HttpExchange` as a neutral annotation to define API interfaces.
- Provide a Spring Cloud OpenFeign like experience for Spring 6.x declarative HTTP clients.
- Support `@RequestMapping` based annotations.
- Not introduce external annotations, easy to migrate to other implementations.


It's definitely worth a try!

**Comment by OlgaMaciaszek** [comment]
We're discussing it within the team now. We will update here once decision's been taken.

**Comment by FilipBehnke** [comment]
@OlgaMaciaszek are there any updates on this topic?

**Comment by OlgaMaciaszek** [comment]
Internal POC planned for this month to discussed with the team. Will post any updates here.

**Comment by XhstormR** [comment]
@OlgaMaciaszek  Since spring-cloud-openfeign is going to enter maintenance mode, will this become the alternative options to openfeign, any update for now?

**Comment by ZaheerUdDeen** [comment]
Hey Folks, any updates here. when the declrative approach will be available natively?

**Comment by OlgaMaciaszek** [comment]
Working on POC at this point. Once that's done we'll able to hold a discussion on adding it to a backlog of a specific release. Will post any relevant updates here.

**Comment by yuexueyang** [comment]
@OlgaMaciaszek Thank you for introduce me to this post. [httpexchange-spring-boot-starter](https://github.com/DanielLiu1123/httpexchange-spring-boot-starter) is really helpful, but may be it still have a small difference from what I want. That's we always configure the base url in database, not in property files, and we need to obtain the base url during every call to the HttpExchange interface to ensure that we use the proper value for each customer (Yes, different customer may use different base url, and the remote site is not constucted by ourself)

**Comment by xtyuns** [comment]
> 不同的客户可能使用不同的基 URL

Maybe you can do it with an `org.springframework.http.client.ClientHttpRequestInterceptor`

**Comment by spencergibb** [comment]
Always get your configuration from the spring environment and use different configuration sources to load from a property or DB or whatever

**Comment by yuexueyang** [comment]
> > 不同的客户可能使用不同的基 URL
> 
> Maybe you can do it with an `org.springframework.http.client.ClientHttpRequestInterceptor`
In my experience, this interceptor can only modify header values, not url.

**Comment by yuexueyang** [comment]
> Always get your configuration from the spring environment and use different configuration sources to load from a property or DB or whatever

Thank you for your advice, what you mean is to configurer  Proxy instance per customer for specified HttpExchange at startup, like Bean instance A for customer A, Bean instance B for customer B, and so on. Is this understanding correct?

**Comment by xtyuns** [comment]
> In my experience, this interceptor can only modify header values, not url.

```kotlin
val urlPrefixedInterceptor = ClientHttpRequestInterceptor { request, body, execution ->
    execution.execute(object : HttpRequest by request {
        override fun getURI(): URI {
            return URI.create("https://example.org${request.uri}")
        }
    }, body)
}
```

**Comment by ch4mpy** [comment]
I ran some [experiments about tooling for `RestClient`, `WebClient` and generated `@HttpExchange` proxies](https://github.com/ch4mpy/spring-addons/tree/master/spring-addons-starter-rest) too. The feature list is not as important as what the `[httpexchange-spring-boot-starter](https://github.com/DanielLiu1123/httpexchange-spring-boot-starter)` I just found above offers, but those experiments forged an opinion on a two things:
- the most important gains are on the underlying client, because it's where the complicated & verbose configuration lies: authorization (`Basic`, `Bearer` from different sources, ...), authentication on HTTP or SOCKS proxy, timeouts, retries, (de)serilizers, etc.
- it's a good thing to de-couple the client bean definition from the generated `@HttpExchange` proxy one. The generated proxy needs a client, not to know how it is configured. So if the generated proxy bean definition contains a client bean reference instead of inlined client configuration, a generated service proxy can use a hand-crafted client (for instance when we want some configuration not provided by the starter we use). Plus several service proxies can use the same client bean.

**Comment by notzero1995** [comment]
> > In my experience, this interceptor can only modify header values, not url.
> 
> You can replace the request, just like the example with kotlin:
> 
> ```kotlin
> val urlPrefixedInterceptor = ClientHttpRequestInterceptor { request, body, execution ->
>     execution.execute(object : HttpRequest by request {
>         override fun getURI(): URI {
>             return URI.create("https://example.org${request.uri}")
>         }
>     }, body)
> }
> ```


example with Java

public class DemoUrlInterceptor implements ClientHttpRequestInterceptor {
    @Override
    public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution) throws IOException {
        HttpRequest newRequest = new HttpRequestWrapper(request) {
            @Override
            public URI getURI() {
                return URI.create("https://example.org" + request.getURI());
            }
        };
        return execution.execute(newRequest, body);
    }

**Comment by notzero1995** [comment]
> > In my experience, this interceptor can only modify header values, not url.
> 
> You can replace the request, just like the example with kotlin:
> 
> ```kotlin
> val urlPrefixedInterceptor = ClientHttpRequestInterceptor { request, body, execution ->
>     execution.execute(object : HttpRequest by request {
>         override fun getURI(): URI {
>             return URI.create("https://example.org${request.uri}")
>         }
>     }, body)
> }
 example with java :
> ```java
    public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution) throws IOException {
        HttpRequest newRequest = new HttpRequestWrapper(request) {
            @Override
            public URI getURI() {
                return URI.create("https://example.org" + request.getURI());
            }
        };
        return execution.execute(newRequest, body);
    }
> ```

**Comment by notzero1995** [comment]
> > In my experience, this interceptor can only modify header values, not url.
> 
> You can replace the request, just like the example with kotlin:
> 
> ```kotlin
> val urlPrefixedInterceptor = ClientHttpRequestInterceptor { request, body, execution ->
>     execution.execute(object : HttpRequest by request {
>         override fun getURI(): URI {
>             return URI.create("https://example.org${request.uri}")
>         }
>     }, body)
> }
> ```
Good approach; sample with Java
```java
public class DemoUrlInterceptor implements ClientHttpRequestInterceptor {
    @Override
    public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution) throws IOException {
        HttpRequest newRequest = new HttpRequestWrapper(request) {
            @Override
            public URI getURI() {
                return URI.create("https://example.org" + request.getURI());
            }
        };
        return execution.execute(newRequest, body);
    }
}
```

**Comment by ch4mpy** [comment]
I ran some [experiments around `@HttpExchange` proxies](https://github.com/ch4mpy/spring-addons/tree/master/spring-addons-starter-rest) too.  The following is [released on maven-central](https://central.sonatype.com/artifact/com.c4-soft.springaddons/spring-addons-starter-rest).

I took a pretty different direction than [httpexchange-spring-boot-starter](https://github.com/DanielLiu1123/httpexchange-spring-boot-starter) for two reasons:
- where I had the most to gain was with the underlying `RestClient` or `WebClient` configuration (mostly requests authorization and HTTP proxy configuration, which can both be cumbersome).
- `@HttpExchange` interfaces are something I usually generate from OpenAPI specs. **Decorating generated code with configuration annotations is not quite an option, reason why I prefer YAML in this case**.

I believe that it's a good thing to de-couple the client bean definition from the generated `@HttpExchange` proxy one. The generated proxy needs a client, not to know how it is configured. So if the generated proxy bean definition contains a client bean reference instead of inlined client configuration:
- **A generated service proxy can use a hand-crafted client** (for instance when we want some configuration not provided by the starter we use).
- Several service proxies can use the same client bean.

I end with YAML for each client and generate no more than pre-configured `RestClient` or `WebClient` beans (or builders if more custom configuration is needed). This is not as magic as the `@FeignClient` auto-detection and implementation, but is great step forward in usability for my use cases. Sample taken from [this repo](https://github.com/ch4mpy/rest-showcase/tree/master/rest-showcase-chouette-api-servlet-addons):
```yaml
com:
  c4-soft:
    springaddons:
      rest:
        client:
          # Exposes a RestClient bean named machinClient (or WebClient in a WebFlux app)
          machin-client:
            base-url: ${machin-api}
            authorization:
              oauth2:
                # Authorize outgoing requests with the Bearer token in the security context (possible only in a resource server app)
                forward-bearer: true
          # Exposes a RestClient.Builder bean named biduleClientBuilder (mind the "expose-builder: true")
          bidule-client:
            base-url: ${bidule-api}
            # Expose the builder instead of an already built client (to fine tune its conf)
            expose-builder: true
            authorization:
              oauth2:
                # Authorize outgoing requests with the Bearer token obtained using an OAuth2 client registration
                oauth2-registration-id: bidule-registration
```
This exposes pre-configured beans that we can auto-wire in any kind of `@Component`, like `@Controller` or `@Service`, or use in `@Configuration`. For instance:
```java
@Configuration
public class RestConfiguration {
  /** 
   * @param machinClient pre-configured by spring-addons-starter-rest using application properties
   * @return a generated implementation of the {@link MachinApi} {@link HttpExchange &#64;HttpExchange}, exposed as a bean named "machinApi".
   */
  @Bean
  MachinApi machinApi(RestClient machinClient) throws Exception {
    return new RestClientHttpExchangeProxyFactoryBean<>(MachinApi.class, machinClient).getObject();
  }

  /** 
   * @param biduleClientBuilder pre-configured using application properties
   * @return a {@link RestClient} bean named "biduleClient"
   */
  @Bean
  RestClient biduleClient(RestClient.Builder biduleClientBuilder) throws Exception {
    // Fine-tune biduleClientBuilder configuration here
    return biduleClientBuilder.build();
  }

  /** 
   * @param biduleClient the bean exposed just above
   * @return a generated implementation of the {@link BiduleApi} {@link HttpExchange &#64;HttpExchange}, exposed as a bean named "biduleApi".
   */
  @Bean
  BiduleApi biduleApi(RestClient biduleClient) throws Exception {
    return new RestClientHttpExchangeProxyFactoryBean<>(BiduleApi.class, biduleClient).getObject();
  }
}
```
This exposes two generated `@HttpExchange` beans named `machinApi` and `biduleApi`, each with a different OAuth2 authorization mechanism (one forwarding the access token in the security context and the other using a client registration with client credentials). And if the `HTTP_PROXY` and `NO_PROXY` environment variables are set, the auto-configured REST clients would use it!

The next step could be replacing the Java code above with something like:
```yaml
        service:
          machin-service:
            client: machinClient
            exchange-class: pf.c4soft.MachinApi
          bidule-service:
            client: biduleClient
            exchange-class: pf.c4soft.BiduleApi
```
But I couldn't find a way to post-process the bean definition registry to do so - and the gains would be marginal.

**Comment by rstoyanchev** [comment]
Thanks for sharing @ch4mpy. 

You mention your approach differs, but as far as I can see httpexchange-spring-boot-starter also [doesn't expect any configuration](https://danielliu1123.github.io/httpexchange-spring-boot-starter/docs/intro#quick-start) on the HTTP interface itself. The various client configuration sets are [configured and associated](https://danielliu1123.github.io/httpexchange-spring-boot-starter/docs/core/configuration) with HTTP interfaces through YAML config.

**Comment by ch4mpy** [comment]
@rstoyanchev Yes, both `httpexchange-spring-boot-starter` and `spring-addons-starter-rest` use some YAML configuration and some of `httpexchange-spring-boot-starter` properties are used for the underlying REST client auto-configuration. But this is where similarities stop.

`httpexchange-spring-boot-starter` aims at auto-detecting `@HttpExchange` interfaces and exposing proxies as beans. There is some flexibility for the underlying REST client (choice of an implementation, timeouts tuning for some of these implementations, and addition of static headers), but too limited for me. It is fine when a client application uses a lot of different `@HttpExchange` interfaces with simple requirements for the underlying REST clients.

In my use cases, REST clients almost always have to authorize their requests, It sometimes have to go through an HTTP proxy requiring some authentication too, and each client app rarely consume that many different `@HttpExchange`. So, I optimized my solution to:
- cover the REST clients auto-configuration I needed: not only timeouts, but also `Basic` or `Bearer` requests authorization, and HTTP proxy. The last two are rather complicated and verbose tasks.
- give complete control over the preconfigured REST client builder for the cases where custom conf is required

If I face use cases where the manual definition of proxies is a problem, maybe will I consider adding some auto-registration for these beans too. For now, the effort is not worth the value I can get out of it.

**Comment by DanielLiu1123** [comment]
Just wanted to share that the [httpexchange-spring-boot-starter](https://github.com/danielliu1123/httpexchange-spring-boot-starter) has released version [3.4.0](https://github.com/DanielLiu1123/httpexchange-spring-boot-starter/releases/tag/v3.4.0), which corresponds to Spring Boot 3.4.0.

This version adds a very interesting new feature: Use manually registered bean if it exists.

If you’re not happy with the autoconfigured HTTP client beans or need more customization, you can manually add the bean using the ["original way"](https://docs.spring.io/spring-framework/reference/integration/rest-clients.html#rest-http-interface):
```java
interface RepositoryService {
    @GetExchange("/repos/{owner}/{repo}")
    Repository getRepository(@PathVariable String owner, @PathVariable String repo);
}

@Configuration
class RepositoryServiceConfiguration {
    @Bean
    public RepositoryService repositoryService(RestClient.Builder restClientBuilder) {
        RestClient restClient = RestClient.builder().baseUrl("https://api.github.com/").build();
        RestClientAdapter adapter = RestClientAdapter.create(restClient);
        HttpServiceProxyFactory factory = HttpServiceProxyFactory.builderFor(adapter).build();
        return factory.createClient(RepositoryService.class);
    }
}
```

You can add a proxy, headers, or anything else you want to customize. When a manually registered HTTP client bean exists, the autoconfigured bean will not be created.

TLDR: You can enjoy the “magic” of autoconfiguration while still having the flexibility to manually add beans using the original approach.

**Comment by OlgaMaciaszek** [comment]
Issue update: the Framework team is working on a DSL to allow registering clients (linked above); then we are planning to add support in Boot, and integrations with Cloud based on that.

**Comment by afattahi54** [comment]
For a new project, I try to use `Http Client` instead of `Feign`.  
I see some projects at which try to make it easier with `Http Client`

- https://danielliu1123.github.io/httpexchange-spring-boot-starter/
- https://github.com/maciejwalkowiak/spring-boot-http-clients

https://github.com/spring-projects/spring-boot/pull/42003
Also it seems that there will be some changes in the `spring` which will be address this issue.

Will there

**Comment by bclozel** [comment]
@afattahi54 yes, see the comment just above yours from Olga.

**Comment by ch4mpy** [comment]
@afattahi54 I suggest that you add [`spring-addons-starter-rest`](https://github.com/ch4mpy/spring-addons/tree/master/spring-addons-starter-rest) to the list of starters you're evaluating. It is more flexible for advanced configuration of the `RestClient`/`WebClient` beans used by the `@HttpExchange` proxies:
- requests authorization:
  - Bearer using a Spring Security OAuth2 client registration
  - Bearer re-using the access token in the security context of an `oauth2ResourceServer` request
  - static header value (API KEY)
  - Basic auth
- base path (property which can be overridden for each deployment)
- auto-applies `HTTP_PROXY` and `NO_PROXY` environment variables (can be overriden with properties)
- configure connection and read timeouts
- enables to choose the  `RestClient` underlying `ClientHttpRequestFactory`, or even to replace it with an hand-crafted one: 
 - `SimpleClientHttpRequestFactory` does not allow `PATCH` requests
 - `JdkClientHttpRequestFactory` is used by default, but it sets headers not supported by some Microsoft middleware
 - `HttpComponentsClientHttpRequestFactory` and `JettyClientHttpRequestFactory` require some additional jars on the classpath
- gives complete flexibility on the `RestClient`/`WebClient` beans configuration: a property allows to expose a pre-configured `Builder`  instead an already built instace to polish the configuration in Java code
- supports many auto-configured `RestClient`/`WebClient` (or builders) beans, each with a name set in properties or derived from the ID in the properties map

**Comment by ch4mpy** [comment]
I recently added the ability to disable SSL certificates validation for a given `RestClient` or `WebClient` (or their builders) auto-configured by [`spring-addons-starter-rest`](https://github.com/ch4mpy/spring-addons/tree/master/spring-addons-starter-rest).

**Comment by ch4mpy** [comment]
@philwebb and @OlgaMaciaszek is there a spec somewhere of what is planned to offer as auto-configuration for the `RestClient` / `WebClient` injected in the generated `@HttpExchanged` proxies?

**Comment by philwebb** [comment]
Commit 5bdb0ecbb4364c8bc773eb3a6369f273b009741d updated the conditions but accidentally had an incorrect `See gh-` number.
